### PR TITLE
Implement timeouts

### LINKED ISSUE #45: Implement Timeouts Submission
After #43 and #44

If a packet is pending, but it is too late to submit it, we should relay a timeout to the original chain (with a non-existence proof from the intended recipient). This is similar to submitting acks, but the queries are a bit different.

### PR DESCRIPTION
**Pull request by willclarktech**
Closes #45

### CODE REVIEW

File: src/lib/utils.ts [hunk 1]
```diff
@@ -289,3 +291,33 @@ export function multiplyCoin({ amount, denom }: Coin, mult: number): Coin {
   const multAmount = Number.parseInt(amount, 10) * mult;
   return { amount: multAmount.toString(), denom };
 }
+
+export function splitPendingPackets(
+  currentHeight: number,
```
Comment (ethanfrey) [inline src/lib/utils.ts:296]: you need to take timeoutTimestamp as well (measured in nanoseconds)

File: src/lib/link.ts [hunk 2]
```diff
@@ -520,6 +521,39 @@ export class Link {
...
+    this.logger.info(`Timeout packets for source ${source}`);
+    console.log('N PACKETS:', packets.length);
+    const { src, dest } = this.getEnds(source);
+
+    // check if we need to update client at all
+    const neededHeight = Math.max(...packets.map((x) => x.height)) + 1;
+    // const headerHeight = await this.updateClientToHeight(source, neededHeight);
+    await this.updateClientToHeight(source, neededHeight);
+
+    const [packet] = packets;
```
Comment (ethanfrey) [inline src/lib/link.ts:537]: Obviously you will need a loop here, but you know that.

Probably just update client.timeoutPacket -> client.timeoutPackets that takes an array of packets with proofs

File: src/lib/link.ts [hunk 3]
```diff
@@ -520,6 +521,39 @@ export class Link {
...
+    const nextSequenceRecvResponse = await dest.client.getNextSequenceRecvProof(
```
Comment (ethanfrey) [inline src/lib/link.ts:538]: You need the same proof query as for submitting an ack. However, when you get it the value should be empty and the proof existing (it is a non-existence proof) - you can console.log the query to ensure this works

File: src/lib/link.ts [hunk 4]
```diff
@@ -520,6 +521,39 @@ export class Link {
     return height;
   }
 
+  public async timeoutPackets(
+    source: Side,
+    packets: readonly PacketWithMetadata[]
+  ): Promise<readonly MsgResult[]> {
```
Comment (ethanfrey) [inline src/lib/link.ts:527]: I wouldn't really return all results. It fails if the tx fails.
Maybe return the block height where they were submitted? Or nothing until you find a need

File: src/lib/link.spec.ts [hunk 5]
```diff
@@ -484,3 +485,85 @@ test.serial(
...
+  // let's make 3 transfer tx at different heights
+  const txHeights = [];
+  for (let i = 0; i < amounts.length; ++i) {
+    const token = { amount: amounts[i].toString(), denom: simapp.denomFee };
+    const { height } = await nodeA.transferTokens(
+      channels.src.portId,
+      channels.src.channelId,
+      token,
+      recipient,
+      timeoutHeights[i]
```
Comment (ethanfrey) [inline src/lib/link.spec.ts:526]: It would be good to make:

1. expired via height
2. non-expired
3. expired via time

File: src/lib/link.spec.ts [hunk 6]
```diff
@@ -484,3 +485,85 @@ test.serial(
...
+  const noPackets = await link.endA.querySentPackets();
+  t.is(noPackets.length, 0);
+
+  // some basic setup for the transfers
+  const recipient = randomAddress(wasmd.prefix);
+  const latestHeight = (await nodeB.latestHeader()).height;
+  const timeoutDestHeight = latestHeight + 2;
+  const submitDestHeight = latestHeight + 500; // valid for 500 blocks
+  const amounts = [1000, 2222, 3456];
+  const timeoutHeights = [
```
Comment (ethanfrey) [inline src/lib/link.spec.ts:511]: This is good for now to ensure splitting works

**Review by ethanfrey — COMMENTED** [review]
Good start, just a few tweaks and this should work

File: src/lib/link.ts (hunk 4 above)
Comment (willclarktech) [reply src/lib/link.ts:527]: Great, I originally returned nothing here, and then switched to this but it's premature optimisation.

**Pull request merged**